src/flamegpu/io/Telemetry.cpp

```diff
@@ -291,5 +303,90 @@ void Telemetry::encourageUsage() {
     }
 }
 
+std::string Telemetry::getConfigDirectory() {
+#ifdef _WIN32
+    char path[MAX_PATH];
+    if (SUCCEEDED(SHGetFolderPathA(NULL, CSIDL_APPDATA, NULL, 0, path))) {
+        return std::string(path);
+    }
+    throw std::runtime_error("Unable to retrieve config directory on Windows");
+#else
+    const char* configHome = std::getenv("XDG_CONFIG_HOME");
+    if (configHome) {
+        return std::string(configHome);
+    }
+    const char* home = std::getenv("HOME");
+    if (home) {
+        return std::string(home) + "/.config";
+    }
+    // try and get the user directory if home is not set
+    struct passwd pwd;
+    struct passwd* result = nullptr;
+    char buffer[4096];
+    int ret = getpwuid_r(getuid(), &pwd, buffer, sizeof(buffer), &result);
+    if (ret == 0 && result != nullptr) {
+        return std::string(pwd.pw_dir) + "/.config";
+    }
+
+    throw std::runtime_error("Unable to retrieve config directory on Linux");
+#endif
+}
+
+std::string Telemetry::generateRandomId() {
+    const char charset[] =
+        "0123456789"
+        "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+        "abcdefghijklmnopqrstuvwxyz";
+    const size_t length = 36;
+    std::random_device rd;
+    std::mt19937 generator(rd());
+    std::uniform_int_distribution<size_t> distribution(0, sizeof(charset) - 2);
+
+    std::string randomId;
+    for (size_t i = 0; i < length; ++i) {
+        randomId += charset[distribution(generator)];
+    }
+    return randomId;
+}
+
+std::string Telemetry::getUserId() {
+    // Generate and a new random 36-character user ID
+    std::string userId = Telemetry::generateRandomId();
+
+    // Try to load an existing ID from file if it exists
+    try {
+        // Determine config file location
+        std::string configDir = Telemetry::getConfigDirectory() + "/FLAMEGPU";
+        std::filesystem::create_directories(configDir);  // Ensure the directory exists
+        std::string filePath = configDir + "/flamegpu_user.cfg";
+
+        // Check if the file exists
+        if (std::filesystem::exists(filePath)) {
+            std::ifstream file(filePath);
+            if (file.is_open()) {
+                std::getline(file, userId);  // overwrite existing Id
+                file.close();
+                return userId;  // Config file and user id found so return it
+            }
+            else {
+                throw std::runtime_error("Unable to open user ID file for reading");
+            }
+        }
```

Comment:
This block doesn't handle someone creating a directory at `configDir/flamegpu_user.cfg` very well.

It doesn't fail expleicitly, but also doesn't emit the warning that should be triggered by the throw.

ends up submitting a blank clientID to telemetry deck" 
\"clientUser\":\"\" 

Again, this is maybe OK but seems like it should submit the random every time version.

Reply:
Logic improved to handle this

Reply:
petty comment, always open file streams as binary. The ascii behaviour, that is windows specific, is dumb and error prone as hell.